management/src/main/java/io/narayana/lra/tools/BrowserCommand.java

```diff
@@ -0,0 +1,420 @@
+
+            currentStoreDir = aa[1];
+
+            return true;
+        }
+
+        return false;
+    }
+
+    public static void main(String[] args) throws Exception {
+        parseArgs(args);
+        BrowserCommand.getCommand(CommandName.START).execute(new PrintStream(System.out, true), null);
+    }
+
+    CommandName name;
+    boolean verbose;
+
+    private BrowserCommand(CommandName name) {
+        this.name = name;
+    }
+
+    abstract void execute(PrintStream printStream, List<String> args) throws Exception;
+
+    protected void help(PrintStream printStream) {
+        if (name.cmdHelp != null)
+            printStream.printf("%s - %s%n", name.name().toLowerCase(), name.cmdHelp);
+    }
+
+    boolean cancel() {return true;}
+
+    private static void setupStore(String storeDir) throws Exception {
+        recoveryPropertyManager.getRecoveryEnvironmentBean().setRecoveryBackoffPeriod(1);
+        Implementations.install();
+        recoveryManager = new RecoveryManagerImple(false);
+        recoveryManager.addModule(new LRARecoveryModule());
+
+        setupStore(storeDir, isHQStore);
+
+        osb = new ObjStoreBrowser();
+        for(String[] typeAndBean: LRA_OSB_TYPES) {
+            osb.addType(typeAndBean[0], typeAndBean[1], typeAndBean[2]);
+        }
+        osb.start();
+    }
+
+    private static void setupStore(String storeDir, boolean hqstore) throws Exception {
+        String storePath = new File(storeDir).getCanonicalPath();
+
+        if (hqstore) {
+            final String storeClassName = com.arjuna.ats.internal.arjuna.objectstore.hornetq.HornetqObjectStoreAdaptor.class.getName();
+
+            BeanPopulator.getDefaultInstance(HornetqJournalEnvironmentBean.class).setStoreDir(storePath);
+            BeanPopulator.getDefaultInstance(ObjectStoreEnvironmentBean.class).setObjectStoreType(storeClassName);
+            BeanPopulator.getNamedInstance(ObjectStoreEnvironmentBean.class, "communicationStore").setObjectStoreType(storeClassName);
+        }
+
+        BeanPopulator.getDefaultInstance(ObjectStoreEnvironmentBean.class).setObjectStoreDir(storePath);
+        BeanPopulator.getNamedInstance(ObjectStoreEnvironmentBean.class, "communicationStore").setObjectStoreDir(storePath);
+        BeanPopulator.getDefaultInstance(CoreEnvironmentBean.class).setNodeIdentifier("no-recovery");
+    }
+
+    private static BrowserCommand[] commands = {
+
+            new BrowserCommand(CommandName.HELP) {
+                @Override
+                void execute(PrintStream printStream, List<String> args) {
+                    for (BrowserCommand command : commands)
+                        command.help(printStream);
+                }
+            },
+
+            new BrowserCommand(CommandName.START) {
+                boolean finished;
+
+                @Override
+                void execute(PrintStream printStream, List<String> args) throws Exception {
+                    setupStore(currentStoreDir);
```

Comment:
Does this require any Narayana that might be interacting with an LRA from this store to be shutdown before starting?

Reply:
This is a standalone browser and designed for offline use but it should be safe to use from the coordinators point of view but the browser may get odd results if the reads interleave with the writes. Note that the file system store uses [locking when it does its writes](https://github.com/jbosstm/narayana/blob/7.1.0.Final/ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/FileSystemStore.java#L373).